lib/cmap/connection.js

```diff
@@ -75,10 +76,14 @@ class Connection extends EventEmitter {
       stream.destroy();
       this.closed = true;
       this[kQueue].forEach(op =>
-        op.cb(new MongoNetworkError(`connection ${this.id} to ${this.address} timed out`))
+        op.cb(
+          new MongoNetworkTimeoutError(`connection ${this.id} to ${this.address} timed out`, {
```

Comment:
So I introduced a new type `MongoNetworkTimeoutError`, and I wanted to see what y'all thought about its usefulness? It helps us here by letting us switch between subtle behaviors in SDAM related to error types, but I could easily have left in the `isNetworkTimeoutError` helper to detect that (or even pass `timeout: true` into the options of the error type here). Is it useful to users that they have a `MongoNetworkError` and a `MongoNetworkTimeoutError`? I kind of lean towards not introducing a new type, but I couldn't quite make myself do it before submitting this. 

/cc @emadum @reggi

Reply:
I think we could definitely go overboard with adding new types for different errors, but I think timeout errors are distinct enough it could be useful. There are some nice shortcuts you can take advantage of when you have different classes for errors, e.g. [filtered catch in bluebird](http://bluebirdjs.com/docs/api/catch.html#filtered-catch).

Reply:
Right, but I guess my question is: would a user ever want to catch a `MongoNetworkTimeoutError` explicitly? Try to imagine doing it yourself, whats an actual use case where I would want to catch a timeout error but not a network error?

Reply:
Definitely seems like an edge case... 

Perhaps on a very spotty network, where a user wants to automatically retry certain operations in case of a timeout error (expected to occur regularly) but log it as a failure for other network errors (which do not occur regularly)?

Reply:
right.. the original helper was:
```
function isNetworkRetryError(err) {
  return err.message.match(/timed out/);
}
```

I think we want to move towards a world where users aren't implementing their own retry logic, but rather will depend on the driver to have robust exponential retry. In my experience most people don't even use the type hierarchy but rather are informed by the text of the error. That might prove the point in both directions: if the message is all that's important then we haven't actually changed anything behaviorally here. Yet, something about introducing a new type often rubs me the wrong way; I increased the API surface we need to support for something that has potentially marginal value to the user.

Reply:
I generally prefer custom error classes rather than having error string-messages or codes, because I think it's more convenient to do an `instanceof` then that sad message match. I'm all for more classes and more specificity.

Reply:
ok, it sounds like consensus is that we stick with the new error type, resolving this comment